# 5. Analyse d'un jeu de données

---

<img src='https://miro.medium.com/max/1200/1*Pl9OyXXrLH_5JiQB8vNx3w.jpeg'>


Dans ce défi, nous explorerons les informations présentes dans une base de données clients.



> **Notre objectif est de visualiser la répartition de ces données.**


## 1. Installation et import de la bibliothèque `pandas`


**Q1**. Dans un premier temps nous installons la bibliothèque `pandas` conçue pour la manipulation de données. Pour cela nous executons dans le terminal la commande suivante :

- **Sur Windows** :

> `pip install pandas`


- **Sur macOS** :

> `pip3 install pandas`


Suivez ensuite les instructions présentes dans le document.

---

**Q2**. Téléchargez le jeu de données à l'adresse [suivante](https://drive.google.com/file/d/13EpnupLYyUBs1Y9iFjBglAvQjh4L5vru/view?usp=sharing) puis placez-le dans le dossier `Code Python`.

---

**Q3**. Importez la bibliothèques `pandas`.

In [ ]:
# Importez pandas avec l'alias pd
import pandas as pd

---

## 2. Data Exploration

Dans cette partie notre objectifs est d'explorer le jeu de données contenant les informations clients.

<img src='https://img.freepik.com/free-vector/business-documents-scanning-electronic-online-doc-with-pie-chart-infographics-data-analytics-annual-report-result-checking-man-with-magnifying-glass-concept-illustration_335657-1770.jpg?w=2000'>

Voici la liste des modules de la bibliotèque pandas qui seront utilisés dans cette partie.

- `pd.read_csv('nom de fichier')` : Pour importer un jeu de données à partir de son 'nom de fichier',
- `df.head(n)` : Pour afficher les n premières lignes du le jeu de données,


- `df.shape()` : Pour afficher les dimensions du jeu de données (nombre de lignes et nombre de colonnes),
- `df.isna().sum()` : Pour afficher le nombre de valeurs manquantes par colonnes


- `df.info()` : Pour afficher les informations du jeu de données,
- `df.describe()` : Pour afficher les données statistiques du jeu de données.


---


**Q4**. Importez votre le jeux de données client dans une variable nomée df. Affichez ensuite les 10 premières lignes de ce jeu de données.

In [19]:
import pandas as pd

df = "data.csv"
data = pd.read_csv(df)

data.head(10)

ID  Gender Ever_Married  Age Graduated     Profession  Work_Experience  \
0  462809    Male           No   22        No     Healthcare              1.0   
1  462643  Female          Yes   38       Yes       Engineer              NaN   
2  466315  Female          Yes   67       Yes       Engineer              1.0   
3  461735    Male          Yes   67       Yes         Lawyer              0.0   
4  462669  Female          Yes   40       Yes  Entertainment              NaN   
5  461319    Male          Yes   56        No         Artist              0.0   
6  460156    Male           No   32       Yes     Healthcare              1.0   
7  464347  Female           No   33       Yes     Healthcare              1.0   
8  465015  Female          Yes   61       Yes       Engineer              0.0   
9  465176  Female          Yes   55       Yes         Artist              1.0   

  Spending_Score  Family_Size  Var_1 Segmentation  
0            Low          4.0  Cat_4            D  
1        Average          3.0  Cat_4            A  
2            Low          1.0  Cat_6            B  
3           High          2.0  Cat_6            B  
4           High          6.0  Cat_6            A  
5        Average          2.0  Cat_6            C  
6            Low          3.0  Cat_6            C  
7            Low          3.0  Cat_6            D  
8            Low          3.0  Cat_7            D  
9        Average          4.0  Cat_6            C

---
**Q5**. Affichez les dimension du jeu de données ainsi que le nombre de valeurs manquantes.

In [23]:
# TODO - Affichez les dimensions du jeu de données à l'aide du module .shape()
import pandas as pd

df = "data.csv"
data = pd.read_csv(df)

data.shape

(8068, 11)

In [24]:
# TODO - Affichez le nombre de données manquantes par colonne a l'aide du module df.isna().sum()
import pandas as pd

df = "data.csv"
data = pd.read_csv(df)

data.isna().sum()

ID                   0
Gender               0
Ever_Married       140
Age                  0
Graduated           78
Profession         124
Work_Experience    829
Spending_Score       0
Family_Size        335
Var_1               76
Segmentation         0
dtype: int64


---
**Q6**. Notez vos premières observations sur l'ensembles des données affichées,
- que contient le jeu de données, quelles sont les variables,
- quelles sont les dimension,
- quelles sont les variables qui contiennent le plus de données manquantes

Le jeu de donnée contient un id, le genre, jamais marié, l'age, la profession, le nombre d'année de travail, le nombre dans la famille,etc. Donc des informations sur de utilisateurs.
Les dimensiosn sont 8068, 11.
Les variables qui contiennent le plus de données manquante sont ever_Married, Graduated,Profession, Work_Experience, Family_Size
Var_A.



---
**Q7**. Affichez les informations puis les données statistics du jeu de données.


In [25]:
# TODO - afichez les statistics du jeu de données avec la méthode df.info()
import pandas as pd

df = "data.csv"
data = pd.read_csv(df)

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8068 entries, 0 to 8067
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               8068 non-null   int64  
 1   Gender           8068 non-null   object 
 2   Ever_Married     7928 non-null   object 
 3   Age              8068 non-null   int64  
 4   Graduated        7990 non-null   object 
 5   Profession       7944 non-null   object 
 6   Work_Experience  7239 non-null   float64
 7   Spending_Score   8068 non-null   object 
 8   Family_Size      7733 non-null   float64
 9   Var_1            7992 non-null   object 
 10  Segmentation     8068 non-null   object 
dtypes: float64(2), int64(2), object(7)
memory usage: 693.5+ KB


In [26]:
# TODO - afichez les statistics du jeu de données avec la méthode df.describe()
import pandas as pd

df = "data.csv"
data = pd.read_csv(df)

data.describe()

ID          Age  Work_Experience  Family_Size
count    8068.000000  8068.000000      7239.000000  7733.000000
mean   463479.214551    43.466906         2.641663     2.850123
std      2595.381232    16.711696         3.406763     1.531413
min    458982.000000    18.000000         0.000000     1.000000
25%    461240.750000    30.000000         0.000000     2.000000
50%    463472.500000    40.000000         1.000000     3.000000
75%    465744.250000    53.000000         4.000000     4.000000
max    467974.000000    89.000000        14.000000     9.000000


---
**Q8**. Notez vos observtions dans la zone de texte ci-dessous :
- Quelles sont les variables qualitatives ?
- Quelles sont les variables quatitatives ?
- Quel est le poids du jeu de données ?
- Que pouvez-vous dire sur l'âge des clients ?
- Que pouvez-vous dire sur la taille des familles ?
- Que pouvez-vous dire sur le nombre d'exprience professionelle ?

## 3. Data cleaning

Dans cette partie nous allons traiter les données présentes dans notre dataframe en remplaçant les données qualitative par des données quantitatives.

<img src='https://miro.medium.com/max/1000/1*ZtO60QvYVzN3rU0VuSSU1g.jpeg'>


Ce traitement de donné nous permetra d'utiliser par la suite nos algorytmes de segmentation et de clustering.

Voici la liste des traitements que nous appliquerons à notre jeu de données :

> 1. Suppressions des lignes contenant des valeurs manquantes,

> 2. Suppressions des colonnes inutiles,

> 3. Remplacement des données quanlitatives par des données quantitatives,

> 4. Sauvegarde du jeu de données après nettoyage.


Voici la liste des modules de pandas qui seront utilisés dans cette partie :
- `.dropna()` : Pour supprimer les lignes contenant des valeurs manquantes dans un jeu de données df. Pour concerver les modifications il faudra réassigner la variable df ==> df = df.dropna()


- `.drop(['nom de la colone'], axis=1)` : Pour supprimer une colonne du jeu de données df. Pour concerver les modifications il faudra réassigner la variable df.


- `df['nom de la colone'].value_counts()` : Pour afficher la répartion des valeurs de la varialbe mentionnée entre crochet.


- `catégorical(df, 'nom de la colone')` : Fonction permettant de remplacer les valeurs qualitatives par des valeurs quantitatives. Pour concerver les modifications il faudra réassigner la variable df ==> df = catégorical(df, 'nom de la colone').


- `df.to_csv('fichier.csv')` : Pour sauvegarder votre jeu de données.


---
**Q9** Exécuter la cellule suivante, elle contient la fonction `catégorical` permettant le traitement de nos données qualitatives.

In [40]:
import pandas as pd

df = pd.read_csv("data.csv")

def catégorical(df, column):
    liste_ = list(df[column].value_counts().index)
    df[column] = df[column].apply(lambda x: liste_.index(x))
    return df

---
**Q10**. Supprimez les lignes contenant des données manquantes puis affichez le jeu de données. Notez ensuite vos remarques dans une zone de texte.

In [39]:
#TODO -  Supprimez les données manquantes avec la méthode .dropna()

import pandas as pd

df = pd.read_csv("data.csv")

def catégorical(df, column):
    liste_ = list(df[column].value_counts().index)
    df[column] = df[column].apply(lambda x: liste_.index(x))
    return df

df.dropna(inplace=True)

result = catégorical(df, 'Age')

print(result)

          ID  Gender Ever_Married  Age Graduated  Profession  Work_Experience  \
0     462809    Male           No   28        No  Healthcare              1.0   
2     466315  Female          Yes   50       Yes    Engineer              1.0   
3     461735    Male          Yes   50       Yes      Lawyer              0.0   
5     461319    Male          Yes   36        No      Artist              0.0   
6     460156    Male           No    9       Yes  Healthcare              1.0   
...      ...     ...          ...  ...       ...         ...              ...   
8062  463002    Male          Yes   13       Yes      Artist              0.0   
8064  464685    Male           No    0        No   Executive              3.0   
8065  465406  Female           No    7       Yes  Healthcare              1.0   
8066  467299  Female           No    3       Yes  Healthcare              1.0   
8067  461879    Male          Yes    2       Yes   Executive              0.0   

     Spending_Score  Family

J4ai supprimé la colonne Age mais ce n'est pas ce qui était demandé sorry XD

---
**Q11**. Supprimez la colonne 'Var_1' ainsi que la colonne 'ID' du jeu de données.


In [41]:
# Suppression de la variable Var_1  avec la méthode .drop(['nom_de_la_colone'], axis=1)

import pandas as pd

df = pd.read_csv("data.csv")
df = df.drop(['Var_1'], axis=1)

print(df)

          ID  Gender Ever_Married  Age Graduated     Profession  \
0     462809    Male           No   22        No     Healthcare   
1     462643  Female          Yes   38       Yes       Engineer   
2     466315  Female          Yes   67       Yes       Engineer   
3     461735    Male          Yes   67       Yes         Lawyer   
4     462669  Female          Yes   40       Yes  Entertainment   
...      ...     ...          ...  ...       ...            ...   
8063  464018    Male           No   22        No            NaN   
8064  464685    Male           No   35        No      Executive   
8065  465406  Female           No   33       Yes     Healthcare   
8066  467299  Female           No   27       Yes     Healthcare   
8067  461879    Male          Yes   37       Yes      Executive   

      Work_Experience Spending_Score  Family_Size Segmentation  
0                 1.0            Low          4.0            D  
1                 NaN        Average          3.0            A  


In [44]:
# Suppression de la variable ID
import pandas as pd

df = pd.read_csv("data.csv")
df = df.drop(['ID'], axis=1)

print(df)

      Gender Ever_Married  Age Graduated     Profession  Work_Experience  \
0       Male           No   22        No     Healthcare              1.0   
1     Female          Yes   38       Yes       Engineer              NaN   
2     Female          Yes   67       Yes       Engineer              1.0   
3       Male          Yes   67       Yes         Lawyer              0.0   
4     Female          Yes   40       Yes  Entertainment              NaN   
...      ...          ...  ...       ...            ...              ...   
8063    Male           No   22        No            NaN              0.0   
8064    Male           No   35        No      Executive              3.0   
8065  Female           No   33       Yes     Healthcare              1.0   
8066  Female           No   27       Yes     Healthcare              1.0   
8067    Male          Yes   37       Yes      Executive              0.0   

     Spending_Score  Family_Size  Var_1 Segmentation  
0               Low          4.0

---
**Q12**. Traitement de la colonne `Gender`

- Affichez la répartition des valeurs de la colonne `Gender`,
- Notez vos observations dans la zonne de texte,
- Appliquez la fonction `catégorial` à la colonne `Gender`,

In [45]:
# Affichez la répartition de la colonne 'Gender' avec la méthode df['nom_de_la_colonne'].value_counts()
import pandas as pd

df = pd.read_csv("data.csv")
gender_distribution = df['Gender'].value_counts()

print(gender_distribution)

df = catégorical(df, 'Gender')

Gender
Male      4417
Female    3651
Name: count, dtype: int64


On voit la répartition Homme/Femme

In [47]:
# Remplacez les valeurs qualitatives par des valeurs quantitatives via la fonction catégorical(df, 'nom_de_la_colonne')
import pandas as pd

df = pd.read_csv("data.csv")
df = catégorical(df, 'Gender')

print(df)

          ID  Gender Ever_Married  Age Graduated     Profession  \
0     462809       0           No   22        No     Healthcare   
1     462643       1          Yes   38       Yes       Engineer   
2     466315       1          Yes   67       Yes       Engineer   
3     461735       0          Yes   67       Yes         Lawyer   
4     462669       1          Yes   40       Yes  Entertainment   
...      ...     ...          ...  ...       ...            ...   
8063  464018       0           No   22        No            NaN   
8064  464685       0           No   35        No      Executive   
8065  465406       1           No   33       Yes     Healthcare   
8066  467299       1           No   27       Yes     Healthcare   
8067  461879       0          Yes   37       Yes      Executive   

      Work_Experience Spending_Score  Family_Size  Var_1 Segmentation  
0                 1.0            Low          4.0  Cat_4            D  
1                 NaN        Average          3.0  

---
**Q13**. Traitement de la colonne `Ever_Married`

- Affichez la répartition des valeurs de la colonne `Ever_Married`,
- Notez vos observations dans la zonne de texte,
- Appliquez la fonction `catégorial` à la colonne `Ever_Married`,

In [54]:
# Affichez la répartition de la colonne 'Ever_Married'
import pandas as pd

df = pd.read_csv("data.csv")
ever_married_distribution = df['Ever_Married'].value_counts()
df['Ever_Married'].fillna(-1, inplace=True)

print(ever_married_distribution)

df = catégorical(df, 'Ever_Married')

Ever_Married
Yes    4643
No     3285
Name: count, dtype: int64


On voit combien sont marié et combien ne le sont pas.

In [56]:
# Remplacez les valeurs qualitatives par des valeurs quantitatives
import pandas as pd

df = pd.read_csv("data.csv")
df = pd.get_dummies(df, columns=['Ever_Married'], drop_first=True)

print(df)

          ID  Gender  Age Graduated     Profession  Work_Experience  \
0     462809    Male   22        No     Healthcare              1.0   
1     462643  Female   38       Yes       Engineer              NaN   
2     466315  Female   67       Yes       Engineer              1.0   
3     461735    Male   67       Yes         Lawyer              0.0   
4     462669  Female   40       Yes  Entertainment              NaN   
...      ...     ...  ...       ...            ...              ...   
8063  464018    Male   22        No            NaN              0.0   
8064  464685    Male   35        No      Executive              3.0   
8065  465406  Female   33       Yes     Healthcare              1.0   
8066  467299  Female   27       Yes     Healthcare              1.0   
8067  461879    Male   37       Yes      Executive              0.0   

     Spending_Score  Family_Size  Var_1 Segmentation  Ever_Married_Yes  
0               Low          4.0  Cat_4            D             False  
1

---
**Q14**. Traitement de la colonne `Graduated`

- Affichez la répartition des valeurs de la colonne `Graduated`,
- Notez vos observations dans la zonne de texte,
- Appliquez la fonction `catégorial` à la colonne `Graduated`,

In [57]:
# Affichez la répartition de la colonne 'Ever_Married'
import pandas as pd

df = pd.read_csv("data.csv")
graduated_distribution = df['Graduated'].value_counts()

print(graduated_distribution)

Graduated
Yes    4968
No     3022
Name: count, dtype: int64


On voit combien on un diplome et combien n'en n'ont pas.

In [65]:
# Remplacez les valeurs qualitatives par des valeurs quantitatives
import pandas as pd

df = pd.read_csv("data.csv")
profession_distribution = df['Profession'].value_counts()

print(profession_distribution)

df = catégorial(df, 'Profession')

Profession
Artist           2516
Healthcare       1332
Entertainment     949
Engineer          699
Doctor            688
Lawyer            623
Executive         599
Marketing         292
Homemaker         246
Name: count, dtype: int64


NameError: name 'catégorial' is not defined

---
**Q15**. Traitement de la colonne `Profession`

- Affichez la répartition des valeurs de la colonne `Profession`,
- Notez vos observations dans la zonne de texte,
- Appliquez la fonction `catégorial` à la colonne `Profession`,

In [66]:
#Affichez la répartition de la colonne 'Profession'
import pandas as pd

df = pd.read_csv("data.csv")
profession_distribution = df['Profession'].value_counts()

print(profession_distribution)

Profession
Artist           2516
Healthcare       1332
Entertainment     949
Engineer          699
Doctor            688
Lawyer            623
Executive         599
Marketing         292
Homemaker         246
Name: count, dtype: int64


On voit le nombre par profession.

In [58]:
# Remplacez les valeurs qualitatives par des valeurs quantitatives
import pandas as pd

df = pd.read_csv("data.csv")
df['Graduated'] = df['Graduated'].map({'No': 0, 'Yes': 1})

print(df)

          ID  Gender Ever_Married  Age  Graduated     Profession  \
0     462809    Male           No   22        0.0     Healthcare   
1     462643  Female          Yes   38        1.0       Engineer   
2     466315  Female          Yes   67        1.0       Engineer   
3     461735    Male          Yes   67        1.0         Lawyer   
4     462669  Female          Yes   40        1.0  Entertainment   
...      ...     ...          ...  ...        ...            ...   
8063  464018    Male           No   22        0.0            NaN   
8064  464685    Male           No   35        0.0      Executive   
8065  465406  Female           No   33        1.0     Healthcare   
8066  467299  Female           No   27        1.0     Healthcare   
8067  461879    Male          Yes   37        1.0      Executive   

      Work_Experience Spending_Score  Family_Size  Var_1 Segmentation  
0                 1.0            Low          4.0  Cat_4            D  
1                 NaN        Average   

---
**Q16**. Traitement de la colonne `Spending_Score`

- Affichez la répartition des valeurs de la colonne `Spending_Score`,
- Notez vos observations dans la zonne de texte,
- Appliquez la fonction `catégorial` à la colonne `Spending_Score`,

In [59]:
# Affichez la répartition de la colonne 'Spending_Score'
import pandas as pd

df = pd.read_csv("data.csv")
spending_score_distribution = df['Spending_Score'].value_counts()

print(spending_score_distribution)

df = catégorical(df, 'Spending_Score')

Spending_Score
Low        4878
Average    1974
High       1216
Name: count, dtype: int64


On voit le nombres de scores par niveau.

In [60]:
# Remplacez les valeurs qualitatives par des valeurs quantitatives
import pandas as pd

df = pd.read_csv("data.csv")
df['Spending_Score'] = df['Spending_Score'].map({'Low': 0, 'Average': 1, 'High': 2})

print(df)

          ID  Gender Ever_Married  Age Graduated     Profession  \
0     462809    Male           No   22        No     Healthcare   
1     462643  Female          Yes   38       Yes       Engineer   
2     466315  Female          Yes   67       Yes       Engineer   
3     461735    Male          Yes   67       Yes         Lawyer   
4     462669  Female          Yes   40       Yes  Entertainment   
...      ...     ...          ...  ...       ...            ...   
8063  464018    Male           No   22        No            NaN   
8064  464685    Male           No   35        No      Executive   
8065  465406  Female           No   33       Yes     Healthcare   
8066  467299  Female           No   27       Yes     Healthcare   
8067  461879    Male          Yes   37       Yes      Executive   

      Work_Experience  Spending_Score  Family_Size  Var_1 Segmentation  
0                 1.0               0          4.0  Cat_4            D  
1                 NaN               1          3.

---
**Q17**. Traitement de la colonne `Segmentation`

- Affichez la répartition des valeurs de la colonne `Segmentation`,
- Notez vos observations dans la zonne de texte,
- Appliquez la fonction `catégorial` à la colonne `Segmentation`,

In [61]:
#TODO - Affichez la répartition de la colonne 'Segmentation'
import pandas as pd

df = pd.read_csv("data.csv")
segmentation_distribution = df['Segmentation'].value_counts()

print(segmentation_distribution)


df = catégorical(df, 'Segmentation')

Segmentation
D    2268
A    1972
C    1970
B    1858
Name: count, dtype: int64


On voit le nombre par segmentation.

In [63]:
# Remplacez les valeurs qualitatives par des valeurs quantitatives
import pandas as pd

df = pd.read_csv("data.csv")
segmentation_mapping = {
    'A': 0,
    'B': 1,
    'C': 2,
    'D': 3
}

df['Segmentation'] = df['Segmentation'].map(segmentation_mapping)

print(df)

          ID  Gender Ever_Married  Age Graduated     Profession  \
0     462809    Male           No   22        No     Healthcare   
1     462643  Female          Yes   38       Yes       Engineer   
2     466315  Female          Yes   67       Yes       Engineer   
3     461735    Male          Yes   67       Yes         Lawyer   
4     462669  Female          Yes   40       Yes  Entertainment   
...      ...     ...          ...  ...       ...            ...   
8063  464018    Male           No   22        No            NaN   
8064  464685    Male           No   35        No      Executive   
8065  465406  Female           No   33       Yes     Healthcare   
8066  467299  Female           No   27       Yes     Healthcare   
8067  461879    Male          Yes   37       Yes      Executive   

      Work_Experience Spending_Score  Family_Size  Var_1  Segmentation  
0                 1.0            Low          4.0  Cat_4             3  
1                 NaN        Average          3.0

---
**Q18**. Sauvegardez votre jeu de données dans un fichier nommé 'data_clean.csv'

In [64]:
# Sauvegardez votre jeu de données avec la méthode df.to_csv('nom_du_fichier.csv')
import pandas as pd

df = pd.read_csv("data.csv")
df.to_csv('data_clean.csv', index=False)